diff --git a/lib/types.d.ts b/lib/types.d.ts
index 816e11e950643e7ee1a0d87fba7d003ec4854e3e..f72373c3fa082c3bed67e068c0719a690749139a 100644
--- a/lib/types.d.ts
+++ b/lib/types.d.ts
@@ -1,10 +1,12 @@
+import type { Awaitable } from "@auth/core/types";
 import type { NextRequest } from "next/server";
+
 /**
  * AppRouteHandlerFnContext is the context that is passed to the handler as the
  * second argument.
  */
 export type AppRouteHandlerFnContext = {
-    params?: Record<string, string | string[]>;
+    params: Awaitable<Record<string, string | string[]>> | Awaitable<undefined>;
 };
 /**
  * Handler function for app routes. If a non-Response value is returned, an error
@@ -14,7 +16,7 @@ export type AppRouteHandlerFn = (
 /**
  * Incoming request object.
  */
-req: NextRequest, 
+req: NextRequest,
 /**
  * Context properties on the request (including the parameters if this was a
  * dynamic route).
